vllm/entrypoints/openai/tools.py

```diff
@@ -0,0 +1,201 @@
+        text = self.template.render_toolcalls(message.tool_calls)
+        if message.content is None:
+            return text
+        else:
+            return message.content + "\n" + text
+
+    def content_from_tool(self, message: ChatCompletionToolMessage) -> str:
+        return self.template.render_toolmessage(message)
+
+    def inject_prompt(self, request: ChatCompletionRequest):
+        """ Generate and inject the prompt for tools calls. """
+        if request.tools is not None and self.call_token_str is not None and len(
+                request.tools):
+            select_tool_choice = request.tool_choice if (
+                request.tool_choice is not None
+                and request.tool_choice != "auto") else None
+            text_inject = self.template.render_toolslist(
+                tool_choice=select_tool_choice, tools_list=request.tools)
+            if isinstance(request.messages, str):
+                request.messages = text_inject + request.messages
+            elif isinstance(request.messages,
+                            List) and len(request.messages) >= 1:
+                request.messages[
+                    0].content = text_inject + request.messages[0].content
+
+
+class ChatPromptCapture:
+
+    def __init__(self):
+        self.content: str = ""
+        self.maybe_function_call = False
+        self.is_function_call = False
+        self.prefix_size = 0
+        self.calls_list: List[{}] = []
+        self.after_new_function_call = False
+
+    def reset(self, reset_calls_list=False):
+        self.content = ""
+        self.maybe_function_call = False
+        self.is_function_call = False
+        self.prefix_size = 0
+        if reset_calls_list:
+            self.calls_list = []
+
+    def num_calls(self):
+        return len(self.calls_list)
+
+    def make_calls_list(self, prompter: OpenAIToolsPrompter):
+        calls_list = self.content.split(prompter.func_call_token())
+        for v_call in calls_list:
+            if len(v_call):
+                try:
+                    call_dict = json.loads(v_call)
+                    if "call" in call_dict:
+                        self.calls_list.append(call_dict)
+                except json.decoder.JSONDecodeError:
+                    # Simply ignore invalid functions calls...
+                    pass
+
+    def to_ChatCompletionMessageToolCall(
+            self, call_id: int) -> Union[ChatCompletionMessageToolCall, None]:
+        if len(self.calls_list) and call_id < len(self.calls_list):
+            call = self.calls_list[call_id]
+            arguments = call["arguments"] if "arguments" in call else None
+            function_call = Function(name=call["call"],
+                                     arguments=json.dumps(arguments)
+                                     if arguments is not None else "")
+            return ChatCompletionMessageToolCall(id="call_" + call["call"] +
```

Comment:
`call['name']` better? 'name' is more general.

Reply:
Done